src/DI/ContainerBuilder.php

```diff
@@ -348,11 +347,37 @@ public function prepareClassList()
 		}
 
 		//  build auto-wiring list
-		$this->classList = [];
+		$this->classList = $preferred = [];
 		foreach ($this->definitions as $name => $def) {
 			if ($class = $def->getImplement() ?: $def->getClass()) {
+				$defAutowired = $def->isAutowired();
+				if (is_array($defAutowired)) {
+					foreach ($defAutowired as $k => $aclass) {
+						if ($aclass === self::THIS_SERVICE) {
+							$defAutowired[$k] = $class;
+						} elseif (!is_a($class, $aclass, TRUE)) {
+							throw new ServiceCreationException("Incompatible class $aclass in autowiring definition of service '$name'.");
+						}
+					}
+				}
+
 				foreach (class_parents($class) + class_implements($class) + [$class] as $parent) {
-					$this->classList[$parent][$def->isAutowired() && empty($this->excludedClasses[$parent])][] = (string) $name;
+					$autowired = $defAutowired && empty($this->excludedClasses[$parent]);
+					if ($autowired && is_array($defAutowired)) {
+						$autowired = FALSE;
+						foreach ($defAutowired as $aclass) {
+							if (is_a($parent, $aclass, TRUE)) {
+								if (empty($preferred[$parent]) && isset($this->classList[$parent][TRUE])) {
+									$this->classList[$parent][FALSE] = array_merge(...$this->classList[$parent]);
+								}
+								$preferred[$parent] = $autowired = TRUE;
+								break;
```

Comment:
You mean that the following should work?

~~~php
test(function () {
	$builder = new DI\ContainerBuilder;
	$builder->addDefinition('bar')
		->setClass('Bar')
		->setAutowired(TRUE);
	$builder->addDefinition('foo')
		->setClass('Foo')
		->setAutowired('Foo');

	Assert::same('foo', $builder->getByType('Foo'));
	Assert::same('bar', $builder->getByType('Bar'));
});
~~~